ciao-sdk/helpers.go

```diff
@@ -0,0 +1,485 @@
+		dumpJSON(body)
+	}
+
+	return nil
+}
+
+func sendCiaoRequest(method string, url string, values []queryValue, body io.Reader, content string) (*http.Response, error) {
+	return sendHTTPRequestToken(method, url, values, scopedToken, body, content)
+}
+
+func getRef(rel string, links []types.Link) string {
+	for _, link := range links {
+		if link.Rel == rel {
+			return link.Href
+		}
+	}
+	return ""
+}
+
+func getCiaoResource(name string, minVersion string) (string, error) {
+	var resources []types.APILink
+	var url string
+
+	if checkPrivilege() {
+		url = buildCiaoURL("")
+	} else {
+		url = buildCiaoURL(fmt.Sprintf("%s", *tenantID))
+	}
+
+	resp, err := sendCiaoRequest("GET", url, nil, nil, "")
+	if err != nil {
+		return "", err
+	}
+
+	err = unmarshalHTTPResponse(resp, &resources)
+	if err != nil {
+		return "", err
+	}
+
+	for _, l := range resources {
+		if l.Rel == name && l.MinVersion == minVersion {
+			return l.Href, nil
+		}
+	}
+
+	return "", errors.New("Supported version of resource not found")
+}
+
+func getCiaoWorkloadsResource() (string, error) {
+	return getCiaoResource("workloads", api.WorkloadsV1)
+}
+
+func checkPrivilege() bool {
+	for i := range tenants {
+		if tenants[i] == "admin" {
+			return true
+		}
+	}
+
+	return false
+}
+
+func limitToString(limit int) string {
+	if limit == -1 {
+		return "Unlimited"
+	}
+
+	return fmt.Sprintf("%d", limit)
+}
+
+func GetCiaoEnvVariables() {
+	controller := os.Getenv(ciaoControllerEnv)
+	ca := os.Getenv(ciaoCACertFileEnv)
+	clientCert := os.Getenv(ciaoClientCertFileEnv)
+
+	infof("Ciao environment variables:\n")
+	infof("\t%s:%s\n", ciaoControllerEnv, controller)
+	infof("\t%s:%s\n", ciaoCACertFileEnv, ca)
+	infof("\t%s:%s\n", ciaoClientCertFileEnv, clientCert)
+
+	if controller != "" && *controllerURL == "" {
+		*controllerURL = controller
+	}
+
+	if ca != "" && *caCertFile == "" {
+		*caCertFile = ca
+	}
+
+	if clientCert != "" && *clientCertFile == "" {
+		*clientCertFile = clientCert
+	}
+}
+
+func CheckCompulsoryOptions() {
+	fatal := ""
+
+	if *clientCertFile == "" {
+		fatal += "Missing required client certificate file\n"
+	}
+	if *controllerURL == "" {
+		fatal += "Missing required Ciao controller URL\n"
+	}
+
+	if fatal != "" {
+		fatalf(fatal)
+	}
+}
+
+func getTenantsFromCertFile(clientCertFile string) ([]string, error) {
+	var certBlock, p *pem.Block
+
+	data, err := ioutil.ReadFile(clientCertFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error loading client cert file")
+	}
+
+	for {
+		p, data = pem.Decode(data)
+		if p == nil {
+			break
+		}
+		if p.Type == "CERTIFICATE" {
+			if certBlock != nil {
+				return nil, errors.Wrap(err, "Incorrect number of certificate blocks in file")
+			}
+			certBlock = p
+		}
+	}
+
+	if certBlock == nil {
+		return nil, errors.New("No certificate block block in cert file")
+	}
+
+	cert, err := x509.ParseCertificate(certBlock.Bytes)
+	if err != nil {
+		return nil, errors.New("Unable to parse x509 certificate data")
```

Comment:
Take a look at the errors package https://github.com/pkg/errors.  This allows you to wrap errors providing a complete context for the error message as it makes its way up the call stack.  We've sort of agreed to use this package in new code and this looks like a good candidate.